sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -501,6 +501,10 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
         case LessThanOrEqual(l, r) => GreaterThan(l, r)
         // not(not(e))  =>  e
         case Not(e) => e
+        // De Morgan's law: !(a || b) => !a && !b
```

Comment:
Yeah, I got it. I meant the same thing, for the normal expression evaluation, this code change probably cause performance regression, it's better to move the CNF stuff into the `predicates.scala` or `patterns.scala`, which only works for the predicate push down.